File: src/main/java/discord/core/ProcessManager.java
package discord.core;

import discord.utils.Utils;

import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

public class ProcessManager {

    private static final ProcessManager INSTANCE = new ProcessManager();

    private final ScheduledExecutorService service = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());

    private final List<Object> services;

    public ProcessManager() {
        services = Utils.getObjects(ProcessManager.class.getPackage().getName() + ".container");
    }

    public static ProcessManager getInstance() {
        return INSTANCE;
    }

    public void init() {
        services.forEach(((service) -> start((ServiceProcessor) service)));
        System.out.println(services.size() + " processes started");
    }

    private void start(final ServiceProcessor processor) {

        service.scheduleAtFixedRate(() -> {
            try {
                processor.init();
            } catch (Exception e) {
                e.printStackTrace();
            }

        }, 0, processor.getTimer().getTime(), processor.getTimer().getUnit());

    }

}

File: src/main/java/discord/core/ProcessManager.kt
package discord.core

import discord.utils.Utils
import java.util.concurrent.Executors
import java.util.function.Consumer

class ProcessManager {

    private val service = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors())
    private val services: List<Any> = Utils.getObjects(ProcessManager::class.java.getPackage().name + ".container")

    fun init() {
        services.forEach(Consumer { service: Any -> start(service as ServiceProcessor) })
        println(services.size.toString() + " processes started")
    }

    private fun start(processor: ServiceProcessor) {
        service.scheduleAtFixedRate({
            try {
                processor.init()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }, processor.delay.toLong(), processor.timer!!.time.toLong(), processor.timer!!.unit)
    }

    companion object {
        val instance = ProcessManager()
    }

}